### ISSUE TITLE
FrankenPHP package

### ISSUE THREAD

**Issue by dunglas** [oerdnj/deb.sury.org#2208]
**Frequently asked questions**

- [x] I have read [Frequently Asked Questions](https://github.com/oerdnj/deb.sury.org/wiki/Frequently-Asked-Questions)

**Is your feature request related to a problem? Please describe.**

Hi,

First of all, thanks for maintaining this packages!

[FrankenPHP](https://frankenphp.dev) creator here. We're considering publishing Debian/Ubuntu packages for FrankenPHP.

Currently, we cannot build a (dynamically built) package that depends on yours because we need ZTS builds (#327).

Would you be open to contributions for adding back ZTS builds?

Also, would you be open to distributing a FrankenPHP package (it's written in Go, so it's quite easy to compile), or would you prefer that we create a separate repository for it?

**Describe the solution you'd like**

Have ZTS and a FrankenPHP packages?

**Describe alternatives you've considered**

1. Building our own packages (but I would prefer not reinventing the wheel)
2. Creating a simple package with the static build of FrankenPHP (but being able to cherry-pick the extensions to install as you allow is better than a big all-in-one binary)

**Distribution (please complete the following information):**
 - OS: all of them?
 - Architecture: all of them?
 - Repository: all of them?

**Package(s) (please complete the following information):**
n/a

**Additional context**
n/a

**Comment by oerdnj** [comment]
> Would you be open to contributions for adding back ZTS builds?

Yes. That said - there might be an alternative to just switch everything to the ZTS builds, but it's hard to come up with any numbers. What is the impact of enabling ZTS builds on the normal Linux machine? I don't know (and I don't really have time to do any measurements here).

> Also, would you be open to distributing a FrankenPHP package (it's written in Go, so it's quite easy to compile)

I am pretty much not eager to take up on the new work with long term maintenance.

> or would you prefer that we create a separate repository for it?

I would prefer that.

**Comment by oerdnj** [comment]
Also - enabling ZTS is by no means something simple. I think that the way to go would be to build both ZTS and no-ZTS from the single `src:phpX.Y` package, but that also means changes to the `php-defaults` package that helps autogenerate all the `d/control` files.

Then all the extensions need to be rebuilt with both ZTS and no-ZTS packages.

Alternative would be to have some kind of "switch" in `php-defaults` and have separate repositories - and just modify the builds accordingly. But that's quite a load on the maintainer. Some of the uploads can be automated I guess, but I never had the time to do this kind of automation.

**Comment by SharkMachine** [comment]
I'm willing to help to make this happen. I don't know how though. I will start at compiling FrankenPHP. Any help is most likely welcome.

**Comment by dunglas** [comment]
@SharkMachine first step would be to enable ZTS by default and do a benchmark to see if it's acceptable or not.

I have the feeling that the overhead will be negligible, especially because people interested in high performance would be able to switch from FPM to FrankenPHP, which is much faster.

Enabling ZTS by default will also have the nice side effect to allow using the parallel extension, which also unlocks nice performance optimizations.

But as pointed out by @oerdnj, we need numbers to be sure it's ok to follow this approach.

**Comment by jorismak** [comment]
So, how can I help? Is there a benchmark to run that someone recommends? 

Is it 'compile php with and without zts' or is it really 'build the ondrej packages with and without zts'?

Im no stranger to builds with autotools and cmake, and have hacked a bit at Debian package configs, but the whole Debian build system is quite new and alien to me .
As in, I can fetch source packages , know some of the dpkg commands to rebuild and can grep and sed through the config...
But by that time mostly I just compile straight from source and leave the Debian packages behind.

So, if there is something I can invest time in and take the load of both ondrej and frankenphp , I'm for it. But I need some pointers to get started then.

**Comment by remicollet** [comment]
FYI, enable ZTS means everything is thread safe.
PHP itself, but all extensions, and all used libraries.

I have tried in the past to continue providing packages (RPM family) with the ZTS engine (CLI only). 
This is a nightmare (mod_php have never relly work in ZTS)
I recently stopped trying this when all the PHP stack was FTBFS (fail to build from sources) in Fedora 41 

I now forgive on this.
So, goog luck if you plan to do this for deb.

**Comment by dunglas** [comment]
@remicollet, thanks to the adoption of FrankenPHP, [almost all extensions are now entirely compatible with ZTS](https://frankenphp.dev/docs/known-issues/#unsupported-php-extensions). That wasn't the case 1 year ago.
A lot of ZTS-related issues (including timeouts, signal handling, etc) also have been fixed in PHP itself.

I don't know if this improved mod_php stability, however.

**Comment by DubbleClick** [comment]
> [@dunglas]
> I have the feeling that the overhead will be negligible, especially because people interested in high performance would be able to switch from FPM to FrankenPHP, which is much faster.

Unfortunately, I think that's only partially relevant. Many webserver management systems like Plesk only work with apache or nginx.

New relic would probably update their extension if it came to it. And then there's php-imap. It's in the process of being deprecated and was already moved out of the main php source, but switching to ZTS everywhere would force users to drop it as well. And imap is definitely not going to get ZTS support...

And then: performance isn't everything, but even then people would likely still want the optimal performance that's within reach. I suspect a lot of users would be upset if php-fpm suddenly drew the short end of the stick.

The in-memory model of FrankenPHP/Swoole really revolutionises php to the point that it makes half of php's performance optimizations superfluous.

**Comment by oerdnj** [comment]
@remicollet Thank you for confirming my experience from the past.

**Comment by DubbleClick** [comment]
> FYI, enable ZTS means everything is thread safe. PHP itself, but all extensions, and all used libraries.
> 
> I have tried in the past to continue providing packages (RPM family) with the ZTS engine (CLI only). This is a nightmare (mod_php have never relly work in ZTS) I recently stopped trying this when all the PHP stack was FTBFS (fail to build from sources) in Fedora 41
> 
> I now forgive on this. So, goog luck if you plan to do this for deb.

https://github.com/crazywhalecc/static-php-cli/issues/624#issuecomment-2714405161

Linking to my comment over at static-php-cli as a potential compromise. Maybe you could share your two cents on the idea.

Thank you *so much* for your work with keeping RHEL based repositiories up-to-date, by the way. Huge fan!

**Comment by rpkoller** [comment]
FrankenPHP is moving under the PHP GitHub organization:  https://externals.io/message/127347

**Comment by MatthiasKuehneEllerhold** [comment]
FrankenPHP now provides .deb packages, but it seems like its packaging php statically into it. So if I want more extensions, I have to compile it myself.
Is there any way @dunglas that frankenphp can just use the system php executable / library with all installed extensions?

Im currently looking to rewrite an app from laminas-mvc to symfony. Switching from apache http2 + php-fpm to frankenphp is easy with symfony but I need a way to install php in a few different configurations. The docker contain seem to have this ability: I can just add extensions to them with `install-php-extensions redis` and frankenphp picks them up.

**Comment by henderkes** [comment]
> Is there any way @dunglas that frankenphp can just use the system php executable / library with all installed extensions?

Only if your system installed version is compiled with ZTS. Which no repository currently does, nor ever did to my knowledge. Remi used to provide ZTS cli `php`, but not `libphp.so`.

I'm currently working on automating a package repository for apk, deb and rpm from static-php. That way frankenphp can link against shared libphp.so. Still a lot of things to figure out, though, even basic ones as the proper location to put it to.

You can manually run static-php-cli with `SPC_CMD_VAR_PHP_EMBED_TYPE="shared"` when you run `bin/spc-gnu-docker build ...`, then you can compile frankenphp from source against that shared libphp.so found in static-php-cli/buildroot/lib.

**Comment by henderkes** [comment]
Small update about a month later, the rpm packages are fully done, but I still need to figure our debian packages and then set up runners to automatically generate them on version change.
